kernel/exec.c

```diff
@@ -6,61 +6,55 @@
 #include "defs.h"
 #include "x86.h"
 #include "elf.h"
+#include "fs.h"
 
 #define FILE_SEPARATOR  '/'
 
-// Find exe path
-static struct inode *find_path(const char *path, char *resultPath, int bufsiz) {
-  struct inode *ip;
+#define EXECVE_USTK_ARGS_NR       (4)
+#define EXECVE_SHEBANG_RESV_ARGS  (2)
+#define EXECVE_SHEBANG_STR       "#!"
+#define EXECVE_SHEBANG_STRLEN     (2)
 
-  if (strchr(path, FILE_SEPARATOR) != 0) {
-    ip = namei(path);
-    if (resultPath != 0)
-      safestrcpy(resultPath, path, bufsiz);
-  } else {
-    // Also find at root path. TODO: Search from $PATH
-    const char PATH[] = "/bin/";
-    const int LEN = sizeof(PATH) - 1;
-    const int BUFSIZ = 128;
-    char exepath[BUFSIZ];
-    memmove(exepath, PATH, LEN);
-    safestrcpy(exepath + LEN, path, BUFSIZ - LEN);
-    ip = namei(exepath);
-    if (resultPath != 0)
-      safestrcpy(resultPath, exepath, bufsiz);
-  }
-  return ip;
-}
+//#define DEBUG_SHOW_EXECVE_ENVS
+//#define DEBUG_SHOW_SHEBANG
 
-// Check ELF header
```

Comment:
After all, I tried to keep functions as it is as far as I can. Would you please believe it.
But to make shebang work correctly,  it will be needed to call execve after shebang processes are completed. To avoid dead locks in such situation, we should lock inode when we need it indeed.
I thought that I had no choice but rewriting codes which you wrote. Sorry.